federation-js/src/composition/utils.ts

```diff
@@ -62,20 +62,39 @@ export function mapFieldNamesToServiceName<Node extends { name: NameNode }>(
 
 export function findDirectivesOnNode(
   node: Maybe<
-    | TypeDefinitionNode
-    | TypeExtensionNode
     | FieldDefinitionNode
-    | SchemaDefinitionNode
+    | FieldNode
+    | DefinitionNode
   >,
   directiveName: string,
 ) {
-  return node && node.directives
+  return node && 'directives' in node && node.directives
     ? node.directives.filter(
         directive => directive.name.value === directiveName,
       )
     : [];
 }
 
+export function findSelectionSetOnNode(
+  node: Maybe<
+    | FieldDefinitionNode
+    | DefinitionNode
```

Comment:
The same `Exclude<>` type I suggested above would give you a similar ergonomic win here.